qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -336,26 +413,36 @@ def _synthesize_annotated_op(self, op: Operation) -> Union[Operation, None]:
         is not an annotated operation).
         """
         if isinstance(op, AnnotatedOperation):
-            # Currently, we ignore the qubits when recursively synthesizing the base operation.
+            # Recursively handle the base operation
+            # This results in QuantumCircuit, DAGCircuit or Gate
             synthesized_op, _ = self._recursively_handle_op(op.base_op, qubits=None)
 
-            # Currently, we depend on recursive synthesis producing either a QuantumCircuit or a Gate.
-            # If in the future we will want to allow HighLevelSynthesis to synthesize, say,
-            # a LinearFunction to a Clifford, we will need to rethink this.
-            if not synthesized_op or not isinstance(synthesized_op, (QuantumCircuit, Gate)):
-                raise TranspilerError(f"HighLevelSynthesis was unable to synthesize {op.base_op}.")
-
             for modifier in op.modifiers:
+
+                # If we have a DAGCircuit at this point, convert it to QuantumCircuit
+                if isinstance(synthesized_op, DAGCircuit):
+                    synthesized_op = dag_to_circuit(synthesized_op, copy_operations=False)
```

Comment:
I look forward to the day when the modifier synthesis can operate on dags and we're not using the circuit and or gate methods. The worst case here we are essentially doing:
```python
circuit_to_dag(dag_to_circuit(synthesized_op).to_gate().control().definition)
```

I don't see a way around it for now, but longer term it'd be good if we didn't do this.

Reply:
Agreed!